src/serac/numerics/petsc_solvers.cpp

```diff
@@ -0,0 +1,1031 @@
+    : PetscPreconditioner(A, prefix)
+{
+  PetscCallAbort(GetComm(), PCSetType(*this, pc_type));
+  PetscCallAbort(GetComm(), PCSetApplicationContext(*this, this));
+  PetscCallAbort(GetComm(), PCSetPreSolve(*this, convertPCPreSolve));
+  clcustom = false;
+  Customize();
+}
+
+PetscPCSolver::PetscPCSolver(MPI_Comm comm_, Operator& op, PCType pc_type, const std::string& prefix)
+    : PetscPreconditioner(comm_, op, prefix)
+{
+  PetscCallAbort(GetComm(), PCSetType(*this, pc_type));
+  PetscCallAbort(GetComm(), PCSetApplicationContext(*this, this));
+  PetscCallAbort(GetComm(), PCSetPreSolve(*this, convertPCPreSolve));
+  clcustom = false;
+  Customize();
+}
+
+void PetscPCSolver::SetOperator(const mfem::Operator& op)
+{
+  PetscBool is_nest;
+  bool      delete_pA = false;
+
+  auto* pA = const_cast<mfem::PetscParMatrix*>(dynamic_cast<const mfem::PetscParMatrix*>(&op));
+  if (!pA) {
+    pA        = new mfem::PetscParMatrix(GetComm(), &op, PETSC_MATAIJ);
+    delete_pA = true;
+  }
+  PetscCallAbort(GetComm(), PetscObjectTypeCompare(*pA, MATNEST, &is_nest));
+  if (!is_nest) {
+    SLIC_DEBUG_ROOT("Not a MATNEST, setting operator directly");
+    serac::logger::flush();
+    fieldsplit_pc_.reset();
+    mfem::PetscPreconditioner::SetOperator(*pA);
+    if (delete_pA) delete pA;
+    return;
+  }
+  SLIC_DEBUG_ROOT("MATNEST detected, setting up fieldsplit");
+
+  if (fieldsplit_pc_) {
+    SLIC_DEBUG_ROOT("Fieldsplit exists, setting operator");
+    fieldsplit_pc_->SetOperator(*pA);
+  } else {
+    SLIC_DEBUG_ROOT("Creating fieldsplit");
+    fieldsplit_pc_ = std::make_unique<mfem::PetscFieldSplitSolver>(GetComm(), *pA, "nest");
+    PetscCallAbort(GetComm(), PCFieldSplitSetType(*fieldsplit_pc_, PC_COMPOSITE_ADDITIVE));
+    PetscCallAbort(GetComm(), PCSetFromOptions(*fieldsplit_pc_));
+    PetscCallAbort(GetComm(), PCSetUp(*fieldsplit_pc_));
+    KSP*     sub_ksps;
+    PetscInt n = 1;
+    PetscCallAbort(GetComm(), PCFieldSplitGetSubKSP(*fieldsplit_pc_, &n, &sub_ksps));
+    PetscCallAbort(GetComm(), KSPSetPC(sub_ksps[0], *this));
+    PC sub_pc_1;
+    PetscCallAbort(GetComm(), KSPGetPC(sub_ksps[1], &sub_pc_1));
+    PetscCallAbort(GetComm(), PCSetType(sub_pc_1, PCNONE));
+    PetscCallAbort(GetComm(), KSPSetFromOptions(sub_ksps[1]));
+    PetscCallAbort(GetComm(), PetscFree(sub_ksps));
+  }
+  Mat A11;
+  PetscCallAbort(GetComm(), MatNestGetSubMat(*pA, 0, 0, &A11));
+  mfem::PetscParMatrix pA11(A11, true);
+  mfem::PetscPreconditioner::SetOperator(pA11);
+  Mat A22;
+  Vec zero;
+  PetscCallAbort(GetComm(), MatNestGetSubMat(*pA, 1, 1, &A22));
+  // Make sure all diagonal elements are set
+  PetscCallAbort(GetComm(), MatSetOption(A22, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE));
+  PetscCallAbort(GetComm(), MatCreateVecs(A22, &zero, NULL));
+  PetscCallAbort(GetComm(), VecSet(zero, 0));
+  PetscCallAbort(GetComm(), MatDiagonalSet(A22, zero, ADD_VALUES));
+  if (delete_pA) delete pA;
```

Comment:
I would prefer not to have single statement if's w/o brackets. It helps prevent common errors that can occur when adding additional lines to a control statement.  With the new clang-format version we recent moved to we can fix these with style...

Reply:
Sounds good, I think this should definitely be enforced by style